File: api/src/services/database/id_generator.py
import time
import threading
import os

class SnowflakeGenerator:
    """
    Snowflake-like ID generator.
    Structure:
    - 1 bit: Unused (always 0)
    - 41 bits: Timestamp in milliseconds
    - 10 bits: Worker/Machine ID
    - 12 bits: Sequence number
    """
    
    # Custom epoch (e.g., project start date: 2026-01-01)
    EPOCH = 1767225600000 
    
    WORKER_ID_BITS = 10
    SEQUENCE_BITS = 12
    
    MAX_WORKER_ID = (1 << WORKER_ID_BITS) - 1
    MAX_SEQUENCE = (1 << SEQUENCE_BITS) - 1
    
    WORKER_ID_SHIFT = SEQUENCE_BITS
    TIMESTAMP_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS
    
    def __init__(self, worker_id: int = 0):
        if worker_id < 0 or worker_id > self.MAX_WORKER_ID:
            # Fallback: use PID or something else if provided worker_id is invalid
            worker_id = os.getpid() & self.MAX_WORKER_ID
            
        self.worker_id = worker_id
        self.sequence = 0
        self.last_timestamp = -1
        self._lock = threading.Lock()

    def _current_timestamp(self) -> int:
        return int(time.time() * 1000)

    def generate(self) -> int:
        with self._lock:
            timestamp = self._current_timestamp()
            
            if timestamp < self.last_timestamp:
                # System clock moved backwards
                timestamp = self.last_timestamp
            
            if timestamp == self.last_timestamp:
                self.sequence = (self.sequence + 1) & self.MAX_SEQUENCE
                if self.sequence == 0:
                    # Sequence exhausted, wait for next millisecond
                    while timestamp <= self.last_timestamp:
                        timestamp = self._current_timestamp()
            else:
                self.sequence = 0
                
            self.last_timestamp = timestamp
            
            # Combine bits
            id = ((timestamp - self.EPOCH) << self.TIMESTAMP_SHIFT) | \
                 (self.worker_id << self.WORKER_ID_SHIFT) | \
                 self.sequence
            
            return id

_generator = SnowflakeGenerator(worker_id=1)

File: equilibra/src/services/idGenerator.ts
/**
 * Snowflake-like ID generator for TypeScript.
 * Uses BigInt to handle 64-bit integers safely.
 */
class SnowflakeGenerator {
  // Custom epoch (Project start date: 2026-01-01)
  private readonly EPOCH = BigInt(1767225600000);

  private readonly WORKER_ID_BITS = BigInt(10);
  private readonly SEQUENCE_BITS = BigInt(12);

  private readonly MAX_WORKER_ID =
    (BigInt(1) << this.WORKER_ID_BITS) - BigInt(1);
  private readonly MAX_SEQUENCE = (BigInt(1) << this.SEQUENCE_BITS) - BigInt(1);

  private readonly WORKER_ID_SHIFT = this.SEQUENCE_BITS;
  private readonly TIMESTAMP_SHIFT = this.SEQUENCE_BITS + this.WORKER_ID_BITS;

  private workerId: bigint;
  private sequence: bigint = BigInt(0);
  private lastTimestamp: bigint = BigInt(-1);

  constructor(workerId: number = 0) {
    const wId = BigInt(workerId);
    if (wId < BigInt(0) || wId > this.MAX_WORKER_ID) {
      this.workerId = BigInt(
        Date.now() % Number(this.MAX_WORKER_ID + BigInt(1)),
      );
    } else {
      this.workerId = wId;
    }
  }

  private currentTimestamp(): bigint {
    return BigInt(Date.now());
  }

  public generate(): bigint {
    let timestamp = this.currentTimestamp();

    if (timestamp < this.lastTimestamp) {
      // Clock moved backwards
      timestamp = this.lastTimestamp;
    }

    if (timestamp === this.lastTimestamp) {
      this.sequence = (this.sequence + BigInt(1)) & this.MAX_SEQUENCE;
      if (this.sequence === BigInt(0)) {
        // Wait for next millisecond
        while (timestamp <= this.lastTimestamp) {
          timestamp = this.currentTimestamp();
        }
      }
    } else {
      this.sequence = BigInt(0);
    }

    this.lastTimestamp = timestamp;

    const id =
      ((timestamp - this.EPOCH) << this.TIMESTAMP_SHIFT) |
      (this.workerId << this.WORKER_ID_SHIFT) |
      this.sequence;

    return id;
  }
}

// Singleton instance
const _generator = new SnowflakeGenerator(1);

/**
 * Generates a unique 64-bit ID as a BigInt.
 */
export const generateId = (): bigint => {
  return _generator.generate();
};

/**
 * Generates a unique 64-bit ID as a string.
 * Recommended for sending to API/Database.
 */
export const generateIdString = (): string => {
  return generateId().toString();
};
